File: packages/tools/definitions/files.ts
import {
  appendFile,
  copyFile,
  createFolder,
  deletePath,
  listFolder,
  moveFile,
  openPath,
  readFile,
  searchFiles,
  writeFile,
} from "../../actions/files";
import type {
  PathArgs,
  SearchFilesArgs,
  TransferArgs,
  WriteFileArgs,
} from "../../shared/types";
import type { ToolDefinition, ToolParameterSchema } from "../types";
import { requireStringField } from "../validation";

function parseTransferArgs(raw: unknown, tool: string): TransferArgs {
  return {
    from: requireStringField(raw, "from", tool),
    to: requireStringField(raw, "to", tool),
  };
}

function transferParameters(verb: string): ToolParameterSchema {
  return {
    type: "object",
    properties: {
      from: {
        type: "string",
        description: `Full or home-relative path of the file to ${verb}.`,
      },
      to: {
        type: "string",
        description:
          "Destination folder or full destination path (inside the home directory).",
      },
    },
    required: ["from", "to"],
    additionalProperties: false,
  };
}

export const searchFilesTool: ToolDefinition<SearchFilesArgs> = {
  name: "search_files",
  description:
    "Search for files and folders by name inside the user's home directory. " +
    "Returns matching full paths (max 20). Use open_path to open a result.",
  parameters: {
    type: "object",
    properties: {
      query: {
        type: "string",
        description: "Part of the file or folder name to look for, e.g. 'resume'.",
      },
    },
    required: ["query"],
    additionalProperties: false,
  },
  parseArgs: (raw) => ({
    query: requireStringField(raw, "query", "search_files"),
  }),
  execute: async ({ query }) => {
    const paths = await searchFiles(query);

    if (paths.length === 0) {
      return `No files matching "${query}" were found.`;
    }

    return `Found ${paths.length} matches:\n${paths.join("\n")}`;
  },
};

export const openPathTool: ToolDefinition<PathArgs> = {
  name: "open_path",
  description:
    "Open a file or folder with its default application. The path must be " +
    "inside the user's home directory; relative paths are resolved from it. " +
    "Common folders like Downloads, Documents, or Desktop work directly.",
  parameters: {
    type: "object",
    properties: {
      path: {
        type: "string",
        description:
          "Full or home-relative path, e.g. 'Downloads' or a result from search_files.",
      },
    },
    required: ["path"],
    additionalProperties: false,
  },
  parseArgs: (raw) => ({ path: requireStringField(raw, "path", "open_path") }),
  execute: ({ path }) => openPath(path),
};

export const moveFileTool: ToolDefinition<TransferArgs> = {
  name: "move_file",
  description:
    "Move a file or folder to a new location inside the home directory. " +
    "Never overwrites; the user is asked to confirm first.",
  parameters: transferParameters("move"),
  parseArgs: (raw) => parseTransferArgs(raw, "move_file"),
  execute: ({ from, to }) => moveFile(from, to),
  confirmQuestion: ({ from, to }) => `Move ${from} to ${to} — yes or no?`,
};

export const copyFileTool: ToolDefinition<TransferArgs> = {
  name: "copy_file",
  description:
    "Copy a file to a new location inside the home directory. Never overwrites.",
  parameters: transferParameters("copy"),
  parseArgs: (raw) => parseTransferArgs(raw, "copy_file"),
  execute: ({ from, to }) => copyFile(from, to),
};

export const deletePathTool: ToolDefinition<PathArgs> = {
  name: "delete_path",
  description:
    "Delete a file or folder by moving it to the recycle bin (recoverable). " +
    "The user is asked to confirm first.",
  parameters: {
    type: "object",
    properties: {
      path: {
        type: "string",
        description: "Full or home-relative path of the file or folder to delete.",
      },
    },
    required: ["path"],
    additionalProperties: false,
  },
  parseArgs: (raw) => ({
    path: requireStringField(raw, "path", "delete_path"),
  }),
  execute: ({ path }) => deletePath(path),
  confirmQuestion: ({ path }) =>
    `Move ${path} to the recycle bin — yes or no?`,
};

export const readFileTool: ToolDefinition<PathArgs> = {
  name: "read_file",
  description:
    "Read the text content of a file inside the home directory (truncated " +
    "for large files). Use to answer questions about a file's contents.",
  parameters: {
    type: "object",
    properties: {
      path: {
        type: "string",
        description: "Full or home-relative path of the text file to read.",
      },
    },
    required: ["path"],
    additionalProperties: false,
  },
  parseArgs: (raw) => ({ path: requireStringField(raw, "path", "read_file") }),
  execute: ({ path }) => readFile(path),
};

export const writeFileTool: ToolDefinition<WriteFileArgs> = {
  name: "write_file",
  description:
    "Create a new text file inside the home directory with the given " +
    "content, e.g. saving a note or a list the user dictated. Never " +
    "overwrites an existing file.",
  parameters: {
    type: "object",
    properties: {
      path: {
        type: "string",
        description:
          "Home-relative or absolute path of the new file, e.g. 'Documents/note.txt'.",
      },
      content: {
        type: "string",
        description: "The text content to write into the file.",
      },
    },
    required: ["path", "content"],
    additionalProperties: false,
  },
  parseArgs: (raw) => ({
    path: requireStringField(raw, "path", "write_file"),
    content: requireStringField(raw, "content", "write_file"),
  }),
  execute: ({ path, content }) => writeFile(path, content),
};

export const appendFileTool: ToolDefinition<WriteFileArgs> = {
  name: "append_file",
  description:
    "Append a line of text to a file, creating the file if it does not " +
    "exist. Use for adding to the user's notes without touching existing " +
    "content.",
  parameters: {
    type: "object",
    properties: {
      path: {
        type: "string",
        description:
          "Home-relative or absolute path of the file, e.g. 'Documents/jarvis-notes.txt'.",
      },
      content: {
        type: "string",
        description: "The text to append as a new line.",
      },
    },
    required: ["path", "content"],
    additionalProperties: false,
  },
  parseArgs: (raw) => ({
    path: requireStringField(raw, "path", "append_file"),
    content: requireStringField(raw, "content", "append_file"),
  }),
  execute: ({ path, content }) => appendFile(path, content),
};

export const listFolderTool: ToolDefinition<PathArgs> = {
  name: "list_folder",
  description:
    "List the files and subfolders inside a folder in the home directory. " +
    "Folder names end with '/'. Use to answer questions like 'what is in " +
    "my Downloads folder'.",
  parameters: {
    type: "object",
    properties: {
      path: {
        type: "string",
        description:
          "Full or home-relative path of the folder, e.g. 'Downloads'.",
      },
    },
    required: ["path"],
    additionalProperties: false,
  },
  parseArgs: (raw) => ({
    path: requireStringField(raw, "path", "list_folder"),
  }),
  execute: async ({ path }) => {
    const entries = await listFolder(path);

    if (entries.length === 0) {
      return `The folder "${path}" is empty.`;
    }

    return `Contents of ${path} (${entries.length} entries):\n${entries.join("\n")}`;
  },
};

export const createFolderTool: ToolDefinition<PathArgs> = {
  name: "create_folder",
  description:
    "Create a new folder inside the user's home directory. Relative paths " +
    "are resolved from the home directory, e.g. 'Documents/Projects/jarvis'.",
  parameters: {
    type: "object",
    properties: {
      path: {
        type: "string",
        description: "Home-relative or absolute path of the folder to create.",
      },
    },
    required: ["path"],
    additionalProperties: false,
  },
  parseArgs: (raw) => ({
    path: requireStringField(raw, "path", "create_folder"),
  }),
  execute: ({ path }) => createFolder(path),
};

File: apps/desktop/src-tauri/src/files.rs
use std::path::{Component, Path, PathBuf};

use log::info;
use walkdir::WalkDir;

const MAX_RESULTS: usize = 20;
const MAX_DEPTH: usize = 6;
const MAX_ENTRIES_SCANNED: usize = 50_000;

/// Directories that are never worth searching and huge enough to matter.
const SKIPPED_DIRS: &[&str] = &[
    "node_modules",
    "target",
    "AppData",
    "Library",
    "dist",
    "build",
    "__pycache__",
    "venv",
];

fn home_dir() -> Result<PathBuf, String> {
    dirs::home_dir().ok_or_else(|| "could not determine the home directory".to_string())
}

/// Resolves a user/AI-supplied path and enforces the home-directory sandbox:
/// relative paths are joined to home, '..' is rejected outright, and absolute
/// paths must already live under home.
fn resolve_in_home(raw: &str) -> Result<PathBuf, String> {
    let path = Path::new(raw.trim());

    if path
        .components()
        .any(|component| matches!(component, Component::ParentDir))
    {
        return Err("paths may not contain '..'".to_string());
    }

    let home = home_dir()?;
    let absolute = if path.is_absolute() {
        path.to_path_buf()
    } else {
        home.join(path)
    };

    if !absolute.starts_with(&home) {
        return Err(format!(
            "access outside the home directory is not allowed: {}",
            absolute.display()
        ));
    }

    Ok(absolute)
}

fn is_searchable(entry: &walkdir::DirEntry) -> bool {
    if !entry.file_type().is_dir() {
        return true;
    }

    match entry.file_name().to_str() {
        Some(name) => !name.starts_with('.') && !SKIPPED_DIRS.contains(&name),
        None => false,
    }
}

#[tauri::command]
pub async fn search_files(query: String) -> Result<Vec<String>, String> {
    let needle = query.trim().to_lowercase();

    if needle.is_empty() {
        return Err("empty search query".to_string());
    }

    info!("search_files called: {}", needle);

    let home = home_dir()?;

    // The walk is blocking work, so keep it off the async runtime's core.
    tauri::async_runtime::spawn_blocking(move || {
        let mut results = Vec::new();
        let mut scanned = 0usize;

        for entry in WalkDir::new(&home)
            .max_depth(MAX_DEPTH)
            .into_iter()
            .filter_entry(is_searchable)
        {
            let Ok(entry) = entry else { continue };

            scanned += 1;
            if scanned > MAX_ENTRIES_SCANNED || results.len() >= MAX_RESULTS {
                break;
            }

            if entry
                .file_name()
                .to_string_lossy()
                .to_lowercase()
                .contains(&needle)
            {
                results.push(entry.path().display().to_string());
            }
        }

        Ok(results)
    })
    .await
    .map_err(|error| format!("file search failed: {}", error))?
}

#[tauri::command]
pub async fn open_path(path: String) -> Result<String, String> {
    info!("open_path called: {}", path);

    let absolute = resolve_in_home(&path)?;

    if !absolute.exists() {
        return Err(format!("path does not exist: {}", absolute.display()));
    }

    open::that_detached(&absolute)
        .map_err(|error| format!("failed to open {}: {}", absolute.display(), error))?;

    Ok(format!("Opened {}", absolute.display()))
}

/// Resolves a destination: if it points at an existing directory the source's
/// file name is appended (so "move X to Documents" does the natural thing).
/// Existing destinations are always rejected — no silent overwrites.
fn resolve_destination(source: &Path, to_raw: &str) -> Result<PathBuf, String> {
    let mut target = resolve_in_home(to_raw)?;

    if target.is_dir() {
        let name = source
            .file_name()
            .ok_or_else(|| "invalid source file name".to_string())?;
        target = target.join(name);
    }

    if target.exists() {
        return Err(format!(
            "destination already exists: {}",
            target.display()
        ));
    }

    match target.parent() {
        Some(parent) if parent.exists() => Ok(target),
        Some(parent) => Err(format!(
            "destination folder does not exist: {}",
            parent.display()
        )),
        None => Err("invalid destination".to_string()),
    }
}

#[tauri::command]
pub async fn move_file(from: String, to: String) -> Result<String, String> {
    info!("move_file called: {} -> {}", from, to);

    let source = resolve_in_home(&from)?;

    if !source.exists() {
        return Err(format!("source does not exist: {}", source.display()));
    }

    let target = resolve_destination(&source, &to)?;

    std::fs::rename(&source, &target).map_err(|error| {
        format!(
            "failed to move {} to {}: {}",
            source.display(),
            target.display(),
            error
        )
    })?;

    Ok(format!("Moved {} to {}", source.display(), target.display()))
}

#[tauri::command]
pub async fn copy_file(from: String, to: String) -> Result<String, String> {
    info!("copy_file called: {} -> {}", from, to);

    let source = resolve_in_home(&from)?;

    if !source.is_file() {
        return Err(format!(
            "source is not a file: {}",
            source.display()
        ));
    }

    let target = resolve_destination(&source, &to)?;

    std::fs::copy(&source, &target).map_err(|error| {
        format!(
            "failed to copy {} to {}: {}",
            source.display(),
            target.display(),
            error
        )
    })?;

    Ok(format!("Copied {} to {}", source.display(), target.display()))
}

/// Deletes by moving to the system recycle bin — never permanent removal.
#[tauri::command]
pub async fn delete_path(path: String) -> Result<String, String> {
    info!("delete_path called: {}", path);

    let absolute = resolve_in_home(&path)?;

    if !absolute.exists() {
        return Err(format!("path does not exist: {}", absolute.display()));
    }

    trash::delete(&absolute)
        .map_err(|error| format!("failed to delete {}: {}", absolute.display(), error))?;

    Ok(format!("Moved {} to the recycle bin", absolute.display()))
}

const MAX_READ_BYTES: u64 = 32_768;

#[tauri::command]
pub async fn read_file(path: String) -> Result<String, String> {
    use std::io::Read;

    info!("read_file called: {}", path);

    let absolute = resolve_in_home(&path)?;

    if !absolute.is_file() {
        return Err(format!("not a file: {}", absolute.display()));
    }

    let file = std::fs::File::open(&absolute)
        .map_err(|error| format!("failed to open {}: {}", absolute.display(), error))?;

    let total_len = file
        .metadata()
        .map(|metadata| metadata.len())
        .unwrap_or(0);

    let mut buffer = Vec::new();
    file.take(MAX_READ_BYTES)
        .read_to_end(&mut buffer)
        .map_err(|error| format!("failed to read {}: {}", absolute.display(), error))?;

    if buffer.contains(&0) {
        return Err(format!(
            "{} looks like a binary file, not text",
            absolute.display()
        ));
    }

    let mut text = String::from_utf8_lossy(&buffer).into_owned();

    if total_len > MAX_READ_BYTES {
        text.push_str("\n...(file truncated)");
    }

    Ok(text)
}

const MAX_WRITE_BYTES: usize = 65_536;

/// Writes a new text file — never overwrites an existing one, mirroring the
/// no-silent-overwrite rule used by move/copy.
#[tauri::command]
pub async fn write_file(path: String, content: String) -> Result<String, String> {
    info!("write_file called: {}", path);

    let absolute = resolve_in_home(&path)?;

    if absolute.exists() {
        return Err(format!("already exists: {}", absolute.display()));
    }

    if content.len() > MAX_WRITE_BYTES {
        return Err("content too large to write".to_string());
    }

    match absolute.parent() {
        Some(parent) if parent.exists() => {}
        Some(parent) => {
            return Err(format!(
                "folder does not exist: {} (create it first with create_folder)",
                parent.display()
            ))
        }
        None => return Err("invalid destination".to_string()),
    }

    std::fs::write(&absolute, content)
        .map_err(|error| format!("failed to write {}: {}", absolute.display(), error))?;

    Ok(format!("Wrote {}", absolute.display()))
}

/// Appends a line of text to a file, creating it if needed. Additive only,
/// so unlike write_file it may target an existing file without confirmation.
#[tauri::command]
pub async fn append_file(path: String, content: String) -> Result<String, String> {
    use std::io::Write;

    info!("append_file called: {}", path);

    let absolute = resolve_in_home(&path)?;

    if content.len() > MAX_WRITE_BYTES {
        return Err("content too large to write".to_string());
    }

    if absolute.is_dir() {
        return Err(format!("is a folder, not a file: {}", absolute.display()));
    }

    match absolute.parent() {
        Some(parent) if parent.exists() => {}
        Some(parent) => {
            return Err(format!(
                "folder does not exist: {} (create it first with create_folder)",
                parent.display()
            ))
        }
        None => return Err("invalid destination".to_string()),
    }

    let mut file = std::fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(&absolute)
        .map_err(|error| format!("failed to open {}: {}", absolute.display(), error))?;

    writeln!(file, "{}", content.trim_end())
        .map_err(|error| format!("failed to append to {}: {}", absolute.display(), error))?;

    Ok(format!("Added to {}", absolute.display()))
}

const MAX_LIST_ENTRIES: usize = 100;

#[tauri::command]
pub async fn list_folder(path: String) -> Result<Vec<String>, String> {
    info!("list_folder called: {}", path);

    let absolute = resolve_in_home(&path)?;

    if !absolute.is_dir() {
        return Err(format!("not a folder: {}", absolute.display()));
    }

    let entries = std::fs::read_dir(&absolute)
        .map_err(|error| format!("failed to list {}: {}", absolute.display(), error))?;

    let mut names: Vec<String> = entries
        .filter_map(|entry| entry.ok())
        .take(MAX_LIST_ENTRIES)
        .map(|entry| {
            let mut name = entry.file_name().to_string_lossy().into_owned();
            if entry.file_type().map(|kind| kind.is_dir()).unwrap_or(false) {
                name.push('/');
            }
            name
        })
        .collect();

    // Folders (trailing '/') sort together, then case-insensitive by name.
    names.sort_by_key(|name| (!name.ends_with('/'), name.to_lowercase()));

    Ok(names)
}

#[tauri::command]
pub async fn create_folder(path: String) -> Result<String, String> {
    info!("create_folder called: {}", path);

    let absolute = resolve_in_home(&path)?;

    if absolute.exists() {
        return Err(format!("already exists: {}", absolute.display()));
    }

    std::fs::create_dir_all(&absolute)
        .map_err(|error| format!("failed to create {}: {}", absolute.display(), error))?;

    Ok(format!("Created folder {}", absolute.display()))
}
